### PR TITLE
Feat: content in sub folders

### PR DESCRIPTION
**Pull request by erezrokah**
Fixes: https://github.com/netlify/netlify-cms/issues/513

~~Still work in progress, though I tested with github/gitlab/bitbucket.~~

~~Enable it by adding `content_in_sub_folders: true` to your folder collection (still need to write the documentation).
By enabling the flag slashes will not be sanitized and will be used for directory structure.
e.g doing:~~

The `slug` option remains the same.
You can add a `path` template for a collection similar to the slug generation template:

```yaml
folder: 'posts'
slug: '{{month}}-{{day}}-{{slug}}'
path: '{{year}}/{{slug}}'
```

Will result in the file saved in `posts/2019/11-19-title.md` for example.

>The `slug` key in the `path` template is the post format value of the slug.

#### TODO:
- [x] Think about filtering items when reading them from the repository
- [x] Think if I really need all the encoding/decoding
- [x] Fix GraphQL query to get nested trees
- [x] Use `path` config instead of `content_in_sub_folders` and keep slug behaviour
- [x] UI improvements
- [x] Test with open authoring
- [X] Only sanitize template variables. **BREAKING CHANGE** - fixes https://github.com/netlify/netlify-cms/issues/1254

#### Things to consider:
- **POSSIBLE BREAKING CHANGE** - folder collections entries are now retrieved up to a depth of 10 for all the backends. If someone relied on the CMS retrieving only up to a depth of 1, they might experience unwanted behavior.
- **Filtering entities on extraction** - feels like a separate issue that can be done once we have more use cases.

#### Further improvements (I think they should be done in another PR):
internally we should not use `slug` in our redux store (and everywhere else for that matter). We should use the entry full path and just URI encode it to make sure it doesn't break the routing.
We should also not reference the term slug in the backends (specifically the GitHub one).

### CODE REVIEW

**Comment by tomrutgers** [comment]
Question: do we need the `content_in_sub_folders` property? The slashes in the slug already imply that right?

**Comment by erezrokah** [comment]
> Question: do we need the `content_in_sub_folders` property? The slashes in the slug already imply that right?

That would make it a breaking change I think

**Comment by chrfritsch** [comment]
Would a site structure like the following also be handled by this?

- index.md
- contact.md
- teams
  - team1.md
  - team2.md
- aboutus.md

**Comment by erezrokah** [comment]
> Would a site structure like the following also be handled by this?
> 
> * index.md
> * contact.md
> * teams
>   
>   * team1.md
>   * team2.md
> * aboutus.md

I think you'll need to set up two collections, one for the `teams` nested directory and one for the parent, then define a filter for each collection (We're still thinking on how to do the filtering as the current mechanism is very limited).

**Comment by chrfritsch** [comment]
Uhh, that's so not nice TBH. Creating a site hierarchy is part of the editors' work (not developers) and should not depend on changing the cms config.

**Comment by erezrokah** [comment]
> Uhh, that's so not nice TBH. Creating a site hierarchy is part of the editors' work (not developers) and should not depend on changing the cms config.

Actually this might be possible with a single collection if you use a field with the slug template.
e.g. `{{fields.slug}}` then have the editor update that field in the UI

**Comment by chrfritsch** [comment]
> > Uhh, that's so not nice TBH. Creating a site hierarchy is part of the editors' work (not developers) and should not depend on changing the cms config.
> 
> Actually this might be possible with a single collection if you use a field with the slug template.
> e.g. `{{fields.slug}}` then have the editor update that field in the UI.
> 
> Another improvement will be to preview the destination path in the UI (like suggested in the issue thread).

Ok, I will give it a try. Thank you. :)

**Comment by barthc** [comment]
@erezrokah  path preview:
Peek 2019-11-24 23-57

And of course  would need some UI improvements.

File: packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControlPane.js [hunk 1]
```diff
@@ -55,6 +56,7 @@ export default class ControlPane extends React.Component {
 
     return (
       <ControlPaneContainer>
+        {collection.has('path') && <PathPreview collection={collection} entry={entry} />}
```
Comment (barthc) [inline packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControlPane.js]: Would be nice if the path preview can come immediately before the `identifier` field.

File: packages/netlify-cms-core/src/selectors/entryDraft.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+import { slugFormatter } from 'Lib/backendHelper';
+
+export const selectDraftPath = (state, collection, entry) => {
+  const config = state.config;
+
+  try {
+    // slugFormatter throws in case an identifier is missing from the entry
+    // we can safely ignore this error as this is just a preview path value
+    return slugFormatter(collection, entry.get('data'), config.get('slug'));
```
Comment (barthc) [inline packages/netlify-cms-core/src/selectors/entryDraft.js]: Would be nice to add the file extension.

File: packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControlPane.js (hunk 1 above)
Comment (erezrokah) [reply packages/netlify-cms-core/src/components/Editor/EditorControlPane/EditorControlPane.js]: Good idea I'll update it

File: packages/netlify-cms-backend-github/src/GraphQLAPI.js [hunk 3]
```diff
@@ -186,6 +186,22 @@ export default class GraphQLAPI extends API {
     }
   }
 
+  addFiles(allFiles, entries, path) {
```
Comment (barthc) [inline packages/netlify-cms-backend-github/src/GraphQLAPI.js]: array reduce would be more cleaner.

Comment (barthc) [reply packages/netlify-cms-backend-github/src/GraphQLAPI.js]: I was thinking something like so:
```
function addFiles(entries, path) {
  return entries.reduce((acc, item) => {
    if (item.type === 'tree') {
      return [...acc, ...addFiles(item.object.entries, `${path}/${item.name}`)];
    }

    return [
      ...acc,
      {
        ...item,
        path: `${path}/${item.name}`,
        size: item.blob && item.blob.size
      }
    ];
  }, []);
}
```

Comment (erezrokah) [reply packages/netlify-cms-backend-github/src/GraphQLAPI.js]: Cool

File: packages/netlify-cms-backend-github/src/queries.js [hunk 4]
```diff
@@ -92,17 +93,47 @@ export const statues = gql`
   ${fragments.object}
 `;
 
+const buildFilesQuery = (depth = 10) => {
+  const PLACE_HOLDER = 'PLACE_HOLDER';
```
Comment (barthc) [inline packages/netlify-cms-backend-github/src/queries.js]: Good thinking with the `PLACE_HOLDER` :+1:

**Comment by erezrokah** [comment]
image

**Comment by erezrokah** [comment]
PathPreview

**Comment by erquhart** [comment]
Two issues with displaying the path:
- most nontechnical users won’t need the info and may be confused by it
- multiple fields could be involved in path formation

What’s more likely to be sought by an editor is the resulting URL path. We could handle this in a separate issue/PR and make some product decisions there.

What do you all think, agree/disagree?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by erezrokah** [comment]
> Two issues with displaying the path:
> 
> * most nontechnical users won’t need the info and may be confused by it
> * multiple fields could be involved in path formation
> 
> What’s more likely to be sought by an editor is the resulting URL path. We could handle this in a separate issue/PR and make some product decisions there.
> 
> What do you all think, agree/disagree?
> 
> Sent with [GitHawk](http://githawk.com)

I agree the preview is very raw in atm. Having the resulting URL path sounds much better

**Comment by erezrokah** [comment]
Removed the path preview completely for now

**Comment by erquhart** [comment]
Actually, second thought, I'm going to hold off on merging since we're going into holiday. If folks here _really_ want to play with it over the next couple of days, make some noise.

**Comment by chrfritsch** [comment]
I think many people, including me, are waiting for this feature. So please publish it 😊💪🏻🔥🚀

**Comment by flyfoxuk** [comment]
Yes please, would like to test it

**Comment by carlmjohnson** [comment]
Can this be turned off? It sort of broke my content organization.

**Comment by hulloanson** [comment]
I have a repo structured like this:

- /
  - blog 1
    - index.md
    - audio.mp3
    - image.jpg
  - blog 2
    - index.md
    - another_audio.mp3
    - more_audio.mp3
    - even_more_audio.mp3

etc...

and the following collection configuration (irrelevant fields not listed):

```yaml
- folder: "/"
  slug: "{{slug}}"
  path: "{{slug}}/index"
  media_folder: ""
  public_folder: ""
```

Is it the intended way to use this feature? I get weird results. Description of what happened below if you're interested. Also any advice on authoring with this structure is greatly appreciated.

------- Here is the experiment results ----------

**Expected behavior:**

Lists entries like "blog 1", "blog 2", etc.

**Actual behavior:**

Produced something like this:

image

Before going any further though, I do understand the empty entries are media files. I probably need to filter them.

The problem is: every time I scrolled to the bottom, all entries with text updated to the same name it happened to load from the last request. Such entries all have the link `https://example.com/#/collections/bites/entries/index

My uneducated guess was the "/index" I wrote in `path` / the "index.md" in each sub-folder messed with redux store keys but I have not studied the implementation... yep. that's all I've tried. just sharing. my concern is still on how I should author with the file structure I proposed.

Thank you!

**Comment by erezrokah** [comment]
Hi @huguestennier, I would suggest moving your content into a dedicated directory, e.g. `content` and then use:
```yaml
- folder: "content"
  slug: "{{slug}}"
  path: "{{slug}}/index"
  media_folder: ""
  public_folder: ""
```

If that doesn't work I suggest opening a new issue for it.

**Comment by hulloanson** [comment]
Thank you @erezrokah ! It works if I put contents into a subfolder :) I don't even have to filter the media files!

**Pull request merged**